### ISSUE TITLE
Inconsistent hover info on GitHub and on Sourcegraph instance

### ISSUE THREAD

**Issue by taras-yemets** [sourcegraph/sourcegraph#47608]
- **Sourcegraph version:**  sourcegraph.com `2a716bd70c29` (built on 2023-02-14)
- **Platform information:**  Sourcegraph browser extension for Chrome `v22.11.24.1820`
- **Pipeline:** [buildkite](https://buildkite.com/sourcegraph/sourcegraph/builds/198593#_)

https://user-images.githubusercontent.com/25318659/218797471-ab7f3117-e9df-4b51-9608-342bc5702232.mov

#### Steps to reproduce:

1. Ensure the browser extension is enabled and it points to sourcegraph.com instance.
2. Hover over [`CallOption`](https://github.com/sourcegraph/jsonrpc2/blob/4fb7cd90793ee6ab445f466b900e6bffb9b63d78/call_opt.go) on GitHub.
3. Click on the Sourcegraph icon next to the file name to open the same file on sourcegraph.com and hover over [`CallOption`](https://sourcegraph.com/github.com/sourcegraph/jsonrpc2@4fb7cd90793ee6ab445f466b900e6bffb9b63d78/-/blob/call_opt.go?L16:29&popover=pinned).

#### Expected behavior:
Code-intel popups on github.com and on sourcegraph.com show the same content.

#### Actual behavior:
"No hover information available" on GitHub and actual `CallOption` hover information on Sourcegraph.

If you would like immediate help on this, please email support@sourcegraph.com (you can still create the issue, but there are no [SLAs](https://about.sourcegraph.com/support/) on issues like there are for support requests).

/cc @felixfbecker

**Comment by philipp-spiess** [comment]
@taras-yemets Have you checked the network requests to see if there are any inconsistencies in the GQL params? I wonder if maybe something really stupid is different like we slice a number off the SHA or so 😅

**Comment by philipp-spiess** [comment]
@taras-yemets Is this with the latest `main` bild of the browser extension or how do I repro?

**Comment by taras-yemets** [comment]
Thanks for looking into it, @philipp-spiess!

 >Have you checked the network requests to see if there are any inconsistencies in the GQL params? I wonder if maybe something really stupid is different like we slice a number off the SHA or so
 
For the mentioned token ([`CallOption`](https://sourcegraph.com/github.com/sourcegraph/jsonrpc2@4fb7cd90793ee6ab445f466b900e6bffb9b63d78/-/blob/call_opt.go?L16:29&popover=pinned)) both browser extension and Sourcegraph web app seem to send similar requests ([`query LocalCodeIntel`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@1969f3e6b69e924b94df38bb3d4af402a0f4c968/-/blob/client/web/src/codeintel/ReferencesPanelQueries.ts?L251-262&subtree=true)) and get similar responses (both including definition). The unexpected behaviour is that despite getting similar responses the browser extension doesn't show the hover information, while web app shows.

- Browser extension:
<details>
<summary>request</summary>

```
{
    "query": "\n    query LocalCodeIntel($repository: String!, $commit: String!, $path: String!) {\n        repository(name: $repository) {\n            commit(rev: $commit) {\n                blob(path: $path) {\n                    localCodeIntel\n                }\n            }\n        }\n    }\n",
    "variables": {
        "repository": "github.com/sourcegraph/jsonrpc2",
        "commit": "4fb7cd90793ee6ab445f466b900e6bffb9b63d78",
        "path": "call_opt.go"
    }
}
```

</details>

<details><summary>response</summary>

```
{
    "data": {
        "repository": {
            "commit": {
                "blob": {
                    "localCodeIntel": "{\"symbols\":[{\"name\":\"c\",\"hover\":\"```go\\nfunc (c callOptionFunc) apply(r *Request) error { return c(r) }\\n```\",\"def\":{\"row\":10,\"column\":6,\"length\":1},\"refs\":[{\"row\":10,\"column\":6,\"length\":1},{\"row\":10,\"column\":57,\"length\":1}]},{\"name\":\"r\",\"hover\":\"```go\\nfunc (c callOptionFunc) apply(r *Request) error { return c(r) }\\n```\",\"def\":{\"row\":10,\"column\":30,\"length\":1},\"refs\":[{\"row\":10,\"column\":30,\"length\":1},{\"row\":10,\"column\":59,\"length\":1}]},{\"name\":\"id\",\"hover\":\"```go\\nfunc PickID(id ID) CallOption {\\n```\\n\\n---\\n\\nPickID returns a call option which sets the ID on a request. Care must be\\ntaken to ensure there are no conflicts with any previously picked ID, nor\\nwith the default sequence ID.\\n\",\"def\":{\"row\":24,\"column\":12,\"length\":2},\"refs\":[{\"row\":24,\"column\":12,\"length\":2},{\"row\":26,\"column\":9,\"length\":2}]},{\"name\":\"meta\",\"hover\":\"```go\\nfunc Meta(meta interface{}) CallOption {\\n```\\n\\n---\\n\\nMeta returns a call option which attaches the given meta object to\\nthe JSON-RPC 2.0 request (this is a Sourcegraph extension to JSON\\nRPC 2.0 for carrying metadata).\\n\",\"def\":{\"row\":15,\"column\":10,\"length\":4},\"refs\":[{\"row\":15,\"column\":10,\"length\":4},{\"row\":17,\"column\":19,\"length\":4}]},{\"name\":\"r\",\"hover\":\"```go\\n\\treturn callOptionFunc(func(r *Request) error {\\n```\",\"def\":{\"row\":16,\"column\":28,\"length\":1},\"refs\":[{\"row\":16,\"column\":28,\"length\":1},{\"row\":17,\"column\":9,\"length\":1}]},{\"name\":\"r\",\"hover\":\"```go\\n\\treturn callOptionFunc(func(r *Request) error {\\n```\",\"def\":{\"row\":25,\"column\":28,\"length\":1},\"refs\":[{\"row\":25,\"column\":28,\"length\":1},{\"row\":26,\"column\":2,\"length\":1}]}]}"
                }
            }
        }
    }
}
```

</details>

- Sourcegraph Web App

<details>
<summary>request</summary>

```
{
    "query": "\n    query LocalCodeIntel($repository: String!, $commit: String!, $path: String!) {\n        repository(name: $repository) {\n            id\n            commit(rev: $commit) {\n                blob(path: $path) {\n                    localCodeIntel\n                }\n            }\n        }\n    }\n",
    "variables": {
        "repository": "github.com/sourcegraph/jsonrpc2",
        "commit": "4fb7cd90793ee6ab445f466b900e6bffb9b63d78",
        "path": "call_opt.go"
    }
}
```

</details>

<details><summary>response</summary>

```
{
    "data": {
        "repository": {
            "id": "UmVwb3NpdG9yeTo0MDk1Mzg=",
            "commit": {
                "blob": {
                    "localCodeIntel": "{\"symbols\":[{\"name\":\"meta\",\"hover\":\"```go\\nfunc Meta(meta interface{}) CallOption {\\n```\\n\\n---\\n\\nMeta returns a call option which attaches the given meta object to\\nthe JSON-RPC 2.0 request (this is a Sourcegraph extension to JSON\\nRPC 2.0 for carrying metadata).\\n\",\"def\":{\"row\":15,\"column\":10,\"length\":4},\"refs\":[{\"row\":15,\"column\":10,\"length\":4},{\"row\":17,\"column\":19,\"length\":4}]},{\"name\":\"r\",\"hover\":\"```go\\n\\treturn callOptionFunc(func(r *Request) error {\\n```\",\"def\":{\"row\":16,\"column\":28,\"length\":1},\"refs\":[{\"row\":16,\"column\":28,\"length\":1},{\"row\":17,\"column\":9,\"length\":1}]},{\"name\":\"id\",\"hover\":\"```go\\nfunc PickID(id ID) CallOption {\\n```\\n\\n---\\n\\nPickID returns a call option which sets the ID on a request. Care must be\\ntaken to ensure there are no conflicts with any previously picked ID, nor\\nwith the default sequence ID.\\n\",\"def\":{\"row\":24,\"column\":12,\"length\":2},\"refs\":[{\"row\":24,\"column\":12,\"length\":2},{\"row\":26,\"column\":9,\"length\":2}]},{\"name\":\"c\",\"hover\":\"```go\\nfunc (c callOptionFunc) apply(r *Request) error { return c(r) }\\n```\",\"def\":{\"row\":10,\"column\":6,\"length\":1},\"refs\":[{\"row\":10,\"column\":6,\"length\":1},{\"row\":10,\"column\":57,\"length\":1}]},{\"name\":\"r\",\"hover\":\"```go\\nfunc (c callOptionFunc) apply(r *Request) error { return c(r) }\\n```\",\"def\":{\"row\":10,\"column\":30,\"length\":1},\"refs\":[{\"row\":10,\"column\":30,\"length\":1},{\"row\":10,\"column\":59,\"length\":1}]},{\"name\":\"r\",\"hover\":\"```go\\n\\treturn callOptionFunc(func(r *Request) error {\\n```\",\"def\":{\"row\":25,\"column\":28,\"length\":1},\"refs\":[{\"row\":26,\"column\":2,\"length\":1},{\"row\":25,\"column\":28,\"length\":1}]}]}"
                }
            }
        }
    }
}
```

</details>

**Comment by taras-yemets** [comment]
> Is this with the latest main build of the browser extension or how do I repro?

@philipp-spiess, it can be reproduced on the latest main build.

**Comment by taras-yemets** [comment]
@philipp-spiess and I were looking into this issue and found out that it is reproduced only with search-based code intelligence. Hover information provided by precise code-intel seems consistent both on the code host and on Sourcegraph instance.


| Code-intel type | CommitID | CodeHost | Sourcegraph |
| -- |-- | -- | -- |
| search-based | [4fb7cd90793ee6ab445f466b900e6bffb9b63d78](https://sourcegraph.com/github.com/sourcegraph/jsonrpc2@4fb7cd90793ee6ab445f466b900e6bffb9b63d78/-/blob/call_opt.go)|<img width="1701" alt="Screenshot 2023-02-17 at 16 32 58" src="https://user-images.githubusercontent.com/25318659/219683376-0c3e0adc-2a88-4b19-9eb9-6d42dd19d839.png">|<img width="1701" alt="Screenshot 2023-02-17 at 16 33 19" src="https://user-images.githubusercontent.com/25318659/219683380-c05149d0-3292-44b4-9806-ec0ac852e96b.png">|
precise | [6864d8cc6d35a79f50745f8990cb4d594a8036f4](https://sourcegraph.com/github.com/sourcegraph/jsonrpc2@6864d8cc6d35a79f50745f8990cb4d594a8036f4/-/blob/call_opt.go?subtree=true)|<img width="1701" alt="Screenshot 2023-02-17 at 16 36 02" src="https://user-images.githubusercontent.com/25318659/219684012-283a1605-bf1f-4980-8024-9d793c924d07.png">|<img width="1701" alt="Screenshot 2023-02-17 at 16 36 09" src="https://user-images.githubusercontent.com/25318659/219684004-499bde7f-ca64-4c78-a508-5a787d14164b.png">|

**Comment by umpox** [comment]
I believe this is also present on the Bitbucket server plugin. 

1. Load: https://bitbucket.sgdev.org/projects/SGDEMO/repos/sourcegraph/browse/cmd/gitserver/main.go
2. Notice that hover information isn't there for search based results
3. Load the same file on https://k8s.sgdev.org/bitbucket.sgdev.org/SGDEMO/sourcegraph@master/-/blob/cmd/gitserver/main.go
4. We have hover information for search based results!

I indexed this file so we had precise code intel and the issue went away here.

Note: This is only on tokens that are cross-bundle (search based intelligence in other files). Search based intel still works for same-file intelligence.

**Comment by umpox** [comment]
I think this is because we fetch same file code intelligence through `stencil` (or some API call). Whereas we run a different query for intelligence in other files.

**Comment by philipp-spiess** [comment]
@umpox Yeah the issue with the browser extension is that we snapshotted the code intel API from the previous code intel extension last july and now it seems to be too outdated already. We have another issue about this as well: https://github.com/sourcegraph/sourcegraph/issues/49199

Seems like this ship is slowly sinking with the different extensions APIs and we need to move all extensions (bext code and ref panel, ideally) to the new code intel API (the newest one is the one we touch here: https://github.com/sourcegraph/sourcegraph/pull/50115)

Might be a bit of work to update the browser extension to use this API but I think it would fix all the inconsistencies we're seeing

**Comment by umpox** [comment]
Narrowed it down to a backend change, not an integrations change. We should be able to fix this without having to release an update to all of our integrations.

This issue only happens when `experimentalFeatures.enableGRPC` is set to `true`.

The `symbolInfo` query no longer returns `null` when symbol information can not be found.

Example `symbolInfo` without gRPC:
```
        "blob": {
          "symbolInfo": null
        }
```

Example `symbolInfo` with gRPC:
```
        "blob": {
          "symbolInfo": {
            "definition": {
              "repo": "",
              "commit": "",
              "path": "",
              "range": null
            },
            "hover": null
          }
        }
```

This change means that [existing code to determine if a valid definition](https://github.com/sourcegraph/code-intel-extensions/blob/fef8b98a81422ca906c47ee82b83a4e24535f1f7/template/src/search/squirrel.ts#L18-L21) is found no longer works, so we no longer fall back to searching for cross-file search-based intelligence (we assume we found a definition).

Interestingly, this is only an issue in our extensions code. The web app has [a check that handles this already](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/shared/src/codeintel/legacy-extensions/search/squirrel.ts?L18%3A1-21%3A10). This was [set 7 months ago in this PR](https://github.com/sourcegraph/sourcegraph/pull/41128). I don't think this check was added in response to any planned changes, presuming we required extra type safety when we inlined the extension code (?). Regardless, it would be good to fix this on the backend to ensure existing integrations still work and ensure we don't need to update all of our integrations to handle this (and check any other code). 

I also confirmed that this would be fixed by moving our integrations code to the newer shared code intel apis, through [this PR](https://github.com/sourcegraph/sourcegraph/pull/50291). That is more of a reason to do that, but that will require additional testing and we should still fix the backend for reasons mentioned above.
